### ISSUE TITLE
Black flicker when alt-tabbing in borderless fullscreen

### ISSUE THREAD

**Issue by NotNite** [godotengine/godot#60665]
### Godot version

v3.4.4.stable.mono.official

### System information

Windows 10, GLES3, AMD Radeon RX 6800 XT

### Issue description

I seem to be experiencing the same behavior described in #11762. The issue was closed several years ago from the author deleting their account, and the comments say to create a new issue, so I've made one.

Setting the following values occurs in a screen flicker when alt-tabbing, even though my assumption is a borderless "fullscreen" game would not do that:
```cs
OS.WindowFullscreen = false;
OS.WindowSize = OS.GetScreenSize();
OS.WindowBorderless = true;
```

Apologies if what I'm saying doesn't make much sense (lol), I'm starting a new project with Godot and the borderless option doesn't seem to do what I expect when the window size is the size of my monitor ("borderless maximized window", as the old issue's title says). This also seems to break screenshots, as described in the old issue I've linked.

### Steps to reproduce

- Create a project
- Set the values above in a script/Godot editor
- Start a project
- Attempt to Alt-Tab
- Notice a small black flicker

**Comment by Calinou** [comment]
This is because Godot actually uses exclusive fullscreen in this situation. You can work this around by setting the window size to be one pixel smaller on one of the axes.

This was remedied in Godot 4.0 by adding separate borderless window and exclusive fullscreen modes, but it's not easy to fix in Godot 3.x.

**Comment by elvisish** [comment]
> This happens because Godot 3.x actually uses exclusive fullscreen if the window size matches the screen size (even if the **Borderless** project setting is enabled). You can work this around by setting the window size to be one pixel smaller on one of the axes.
> 
> This was remedied in Godot 4.0alpha by adding separate borderless window and exclusive fullscreen modes. Unfortunately, it's not easy to fix in Godot 3.x which doesn't use the backwards-incompatible [DisplayServer abstraction](https://godotengine.org/article/core-refactoring-progress-report-1).

I was sure that Juan had mentioned exclusive fullscreen being unnecessary on Windows 10 and upwards, but I've always wished 3.x had exclusive fullscreen as it can definitely boost performance. Is it back-portable?

**Comment by Calinou** [comment]
> Is it back-portable?

No, as it needs to be redone from scratch for `3.x` since `master` uses an entirely different windowing implementation.

**Comment by lekoder** [comment]
The engine should use exclusive fullscreen only if the application demands it, not with some magic window sizes. You assume here that there are no side-effects of supplementing screen-matching borderless window with exclusive fullscreen - but there are, and games implement the choice for the player to use one or the other as a troubleshooting tool. Having to hack my way around to have screen-covering borderless window instead of full-screen is hardly developer-friendly, especially if this is as obfuscated as it is here.

**Comment by FoxLeader** [comment]
Since I stumbled on the same issue and this is how I did it without magic window sizes:

        auto os = OS::get_singleton();
        Vector2 windowSize = os->get_screen_size();
        HWND winHandle = reinterpret_cast<HWND>(os->get_native_handle(OS::HandleType::WINDOW_HANDLE));
        LONG winStyle = GetWindowLong(winHandle, GWL_STYLE);
        winStyle &= ~(WS_CAPTION | WS_THICKFRAME | WS_MINIMIZEBOX | WS_MAXIMIZEBOX | WS_SYSMENU);
        SetWindowLongPtr(winHandle, GWL_STYLE, winStyle);

        LONG winExStyle = GetWindowLong(winHandle, GWL_EXSTYLE);
        winExStyle &= ~(WS_EX_DLGMODALFRAME | WS_EX_CLIENTEDGE | WS_EX_STATICEDGE);
        SetWindowLongPtr(winHandle, GWL_EXSTYLE, winExStyle);

        SetWindowPos(winHandle, HWND_TOP, 0, 0, windowSize.x, windowSize.y, SWP_NOSENDCHANGING);

The key here is the SWP_NOSENDCHANGING parameter, this will avoid notifying Godot about the window resize (but the OGL framebuffer will update correctly as far as I can tell).

I'm not very familiar with the Engine Source Code, but if anyone knows about side effects please let me know.
Also, if anyone wants to add this as a workaround for the 3.x branch, please do!

**Comment by FoxLeader** [comment]
Side note: I found that if you constrain the mouse cursor BEFORE setting your window borderless using the code above, Godot won't update the constrained area, but you can call set_mouse_mode(Input::MOUSE_MODE_CONFINED); after SetWindowPos() to force Godot to update.

**Comment by gubblesdev** [comment]
I can confirm that despite the DisplayServer overhaul in Godot 4.0 (stable release), this is still equivalent to the behavior I'm seeing in Windows 10. Should this issue here be updated to add 4.0 as well, or should I submit that as a new issue?

**Comment by Calinou** [comment]
> I can confirm that despite the DisplayServer overhaul in Godot 4.0 (stable release), this is still equivalent to the behavior I'm seeing in Windows 10. Should this issue here be updated to add 4.0 as well, or should I submit that as a new issue?

Godot 4 has separate Fullscreen and Exclusive Fullscreen window modes you can choose in the advanced Project Settings.

**Comment by pegasusearl** [comment]
Happened on Godot 3.5.3. Trying to use Display Capture on OBS, it can only record one frame of the game after splash screen.

> You can work this around by setting the window size to be one pixel smaller on one of the axes.

I made this comment because this trick doesn't work or did I misunderstand the trick?
My screen size is 1920x1080 and this is my setting
![image](https://github.com/user-attachments/assets/867a67a5-9970-4697-a942-f6a394998225)

**Comment by Knand1861** [comment]
![image](https://github.com/user-attachments/assets/58e173f2-0053-4148-ab64-1854166f39ad)
If you subtract one pixel it's technically is not fullscreen anymore and it works 🤓

**Comment by lekoder** [comment]
I favour a workaround of adding `Vector2(0,1)`. Almost unnoticeable.

**Comment by Imberflur** [comment]
Is this fixed by https://github.com/godotengine/godot/pull/107583?

**Comment by Calinou** [comment]
Fixed by https://github.com/godotengine/godot/pull/107583 (after enabling the `display/window/size/use_nonexclusive_fullscreen` project setting), closing.
